### PR TITLE
GEOMESA-708 Expose BatchWriterConfig

### PR DESCRIPTION
**Pull request by jnh5y**
* GeoMesaBatchWriterConfig looks for various settings on the classpath and uses them when creating BatchWriterConfigs.

### CODE REVIEW

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  val DEFAULTLATENCY   = 10000l   // 10 seconds
+  val DEFAULTMAXMEMORY = 1000000l // 1 megabyte
+  val DEFAULTTHREADS   = 10
+
+  protected [util] def fetchProperty(prop: String): Option[Long] =
+    for {
+      p <- Option(System.getProperty(prop))
+      l <- Try(java.lang.Long.parseLong(p)).toOption
+    } yield l
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: slightly confusing 1/l here, but probably not worth changing.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.util.concurrent.TimeUnit
+
+import com.typesafe.scalalogging.slf4j.Logging
+import org.apache.accumulo.core.client.BatchWriterConfig
+
+import scala.util.Try
+
+object GeoMesaBatchWriterConfig extends Logging {
+  val WRITERLATENCYSECONDS = "geomesa.batchwriter.latency.seconds"  // Measured in millis
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: comment should be seconds i assume

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    val bwc = new BatchWriterConfig
+
+    fetchProperty(WRITERLATENCYSECONDS) match {
+      case Some(latency) =>
+        logger.trace(s"GeoMesaBatchWriter config: maxLatency set to $latency seconds.")
+        bwc.setMaxLatency(latency, TimeUnit.SECONDS)
+      case None =>
+        val milliLatency = fetchProperty(WRITERLATENCYMILLIS).getOrElse(DEFAULTLATENCY)
+        logger.trace(s"GeoMesaBatchWriter config: maxLatency set to $milliLatency milliseconds.")
+        bwc.setMaxLatency(milliLatency, TimeUnit.MILLISECONDS)
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaBatchWriterConfig.scala]: might just want to simplify this and make all the properties in milliseconds.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala [hunk 4]
```diff
@@ -121,7 +122,7 @@ class AccumuloDataStore(val connector: Connector,
 
   // TODO memory should be configurable
   private val defaultBWConfig =
-    new BatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(writeThreads)
+    GeoMesaBatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(writeThreads)
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala]: you can probably get rid of the 'set max memory' here and remove the TODO above.

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/GeoMesaMetadata.scala [hunk 5]
```diff
@@ -57,7 +57,7 @@ class AccumuloBackedMetadata(connector: Connector,
 
   // TODO memory should be configurable
   private val metadataBWConfig =
-    new BatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(1)
+    GeoMesaBatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(1)
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/GeoMesaMetadata.scala]: ditto about max memory

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/stats/StatWriter.scala [hunk 6]
```diff
@@ -60,7 +61,7 @@ object StatWriter extends Runnable with Logging {
 
   private val writeDelayMillis = 1000
 
-  private val batchWriterConfig = new BatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(5)
+  private val batchWriterConfig = GeoMesaBatchWriterConfig().setMaxMemory(10000L).setMaxWriteThreads(5)
```
Comment (elahrvivaz) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/stats/StatWriter.scala]: ditto max memory... although now that i think about it, maybe we'd want to cap stat writer and metadata writers at lower amounts of memory, and have the config only affect the main geomesa feature writing.

**Comment by elahrvivaz** [comment]
Looks good. As a side note, as part of time first indices I created a geomesa.conf that could be included on the classpath to override certain defaults. At some point we might want to take that approach as the number of things we want to configure grows.
See:
https://github.com/locationtech/geomesa/blob/9a122dc58726c40a69c8522b31c5be68cadd3fc8/geomesa-core/src/main/resources/org/locationtech/geomesa/core/util/geomesa.conf
https://github.com/locationtech/geomesa/blob/9a122dc58726c40a69c8522b31c5be68cadd3fc8/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/GeoMesaConfig.scala
https://github.com/locationtech/geomesa/blob/9a122dc58726c40a69c8522b31c5be68cadd3fc8/geomesa-core/src/main/scala/org/locationtech/geomesa/core/util/GeoMesaCoreConfig.scala

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala (hunk 4 above)
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/data/AccumuloDataStore.scala]: Good catch!

File: geomesa-core/src/main/scala/org/locationtech/geomesa/core/stats/StatWriter.scala (hunk 6 above)
Comment (jnh5y) [inline geomesa-core/src/main/scala/org/locationtech/geomesa/core/stats/StatWriter.scala]: Maybe there should be GeoMesaBatchWriter.get{Feature,Stats,Metadata}Config methods?  (The last two could share.)

**Comment by elahrvivaz** [comment]
Closing in favor of https://github.com/locationtech/geomesa/pull/510

**Pull request closed without merging**